pkg/controller/machine_util.go

```diff
@@ -266,3 +268,121 @@ func nodeConditionsHaveChanged(machineConditions []v1.NodeCondition, nodeConditi
 
 	return false
 }
+
+// syncMachineNodeTemplate syncs nodeTemplates between machine and corresponding node-object.
+// It ensures, that any nodeTemplate element available on Machine should be available on node-object.
+// Although there could be more elements already available on node-object which will not be touched.
+func (c *controller) syncMachineNodeTemplates(machine *v1alpha1.Machine) error {
+	if machine.Status.Node == "" {
+		return fmt.Errorf("Error: Node field is empty on Machine-object, %q", machine.Name)
+	}
+	node, err := c.nodeLister.Get(machine.Status.Node)
+	if err != nil || node == nil {
+		glog.Errorf("Error: Could not get the node-object or node-object is missing - err: %q", err)
+		return err
+	}
+	nodeCopy := node.DeepCopy()
+
+	// Sync Labels
+	labelsChanged := SyncMachineLabels(machine, nodeCopy)
+
+	// Sync Annotations
+	annotationsChanged := SyncMachineAnnotations(machine, nodeCopy)
+
+	// Sync Taints
+	taintsChanged := SyncMachineTaints(machine, nodeCopy)
+
+	// Update node-object with latest nodeTemplate elements if elements have changed.
+	if labelsChanged || annotationsChanged || taintsChanged {
+		_, err := c.targetCoreClient.Core().Nodes().Update(nodeCopy)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SyncMachineLabels syncs the labels of the machine with node-objects.
+// It returns true if update is needed else false.
+func SyncMachineLabels(machine *v1alpha1.Machine, node *v1.Node) bool {
+	mCopy, nCopy := machine.Spec.NodeTemplateSpec.Labels, node.Labels
+	updateNeeded := false
+
+	if mCopy == nil {
+		return false
+	}
+	if nCopy == nil {
+		nCopy = make(map[string]string)
+	}
+
+	for mkey, mvalue := range mCopy {
+		// if key doesnt exists or value does not match
+		if _, ok := nCopy[mkey]; !ok || mvalue != nCopy[mkey] {
+			updateNeeded = true
+		}
+		nCopy[mkey] = mvalue
```

Comment:
Here `nCopy` is not really a copy, is it? It will actually modify the `node.Labels`? Is this expected? According to [this line], it seems that `nCopy` is treated as a true (possibly shallow) copy.